File: internal/ws/protocol.go
package ws

import "encoding/json"

// ClientMessage is sent from the browser to the server.
// If ID is non-nil, the client expects an ack response with the same ID.
type ClientMessage struct {
    ID    *int64          `json:"id,omitempty"`
    Event string          `json:"event"`
    Args  json.RawMessage `json:"args"`
}

// AckMessage is sent from the server to the client in response to a request with an ID.
type AckMessage[T any] struct {
    ID   int64 `json:"id"`
    Data T     `json:"data"`
}

// ServerMessage is a server-initiated push (no ack expected).
// Data holds the event payload (a single value of any type).
type ServerMessage[T any] struct {
    Event string `json:"event"`
    Data  T      `json:"data"`
}

// OkResponse is the standard ack payload for successful operations.
type OkResponse struct {
    OK    bool   `json:"ok"`
    Msg   string `json:"msg,omitempty"`
    Token string `json:"token,omitempty"`
}

// ErrorResponse is the standard ack payload for failed operations.
type ErrorResponse struct {
    OK      bool   `json:"ok"`
    Msg     string `json:"msg"`
    MsgI18n bool   `json:"msgi18n,omitempty"`
}

// TerminalJoinArgs is the payload for "terminalJoin" events.
type TerminalJoinArgs struct {
    Type      string `json:"type"`
    Stack     string `json:"stack,omitempty"`
    Service   string `json:"service,omitempty"`
    Container string `json:"container,omitempty"`
    Shell     string `json:"shell,omitempty"`
}

// TerminalJoinResponse is the ack payload for "terminalJoin".
type TerminalJoinResponse struct {
    OK        bool   `json:"ok"`
    SessionID uint16 `json:"sessionId"`
    Msg       string `json:"msg,omitempty"`
}

// TerminalLeaveArgs is the payload for "terminalLeave" events.
type TerminalLeaveArgs struct {
    SessionID uint16 `json:"sessionId"`
}

// TerminalExitedData is the payload for "terminalExited" server push events.
type TerminalExitedData struct {
    SessionID uint16 `json:"sessionId"`
}

File: backend-rust/src/ws/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::value::RawValue;

/// Client-to-server message.
#[derive(Deserialize, Debug)]
pub struct ClientMessage {
    pub id: Option<i64>,
    pub event: String,
    #[serde(default)]
    pub args: Option<Box<RawValue>>,
}

/// Server-to-client ack (response to a request with an id).
#[derive(Serialize)]
pub struct AckMessage<T: Serialize> {
    pub id: i64,
    pub data: T,
}

/// Server-to-client push event (unsolicited).
#[derive(Serialize)]
pub struct ServerMessage<T: Serialize> {
    pub event: String,
    pub data: T,
}

/// Common ack payload for success.
#[derive(Serialize)]
pub struct OkResponse {
    pub ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub msg: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token: Option<String>,
}

/// Ack payload for terminal join responses.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionResponse {
    pub ok: bool,
    pub session_id: u16,
}

/// Common ack payload for errors.
#[derive(Serialize)]
pub struct ErrorResponse {
    pub ok: bool,
    pub msg: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub msgi18n: Option<bool>,
}

impl OkResponse {
    pub fn simple() -> Self {
        Self {
            ok: true,
            msg: None,
            token: None,
        }
    }
}

/// Generic wrapper for broadcast events with an `items` map.
/// Used by stacks, containers, networks, images, and volumes broadcasts.
#[derive(Serialize)]
pub struct ItemsEvent<T: Serialize> {
    pub items: T,
}

impl ErrorResponse {
    pub fn new(msg: impl Into<String>) -> Self {
        Self {
            ok: false,
            msg: msg.into(),
            msgi18n: None,
        }
    }

    pub fn i18n(msg: impl Into<String>) -> Self {
        Self {
            ok: false,
            msg: msg.into(),
            msgi18n: Some(true),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // ── OkResponse ──────────────────────────────────────────────────────

    #[test]
    fn ok_response_simple() {
        let r = OkResponse::simple();
        assert!(r.ok);
        assert!(r.msg.is_none());
        assert!(r.token.is_none());
    }

    #[test]
    fn ok_response_omits_optional_fields() {
        let r = OkResponse::simple();
        let json = serde_json::to_string(&r).unwrap();
        assert!(!json.contains("msg"));
        assert!(!json.contains("token"));
        assert!(json.contains("\"ok\":true"));
    }

    // ── ErrorResponse ───────────────────────────────────────────────────

    #[test]
    fn error_response_new() {
        let r = ErrorResponse::new("something failed");
        assert!(!r.ok);
        assert_eq!(r.msg, "something failed");
        assert!(r.msgi18n.is_none());
    }

    #[test]
    fn error_response_i18n() {
        let r = ErrorResponse::i18n("authIncorrectCreds");
        assert!(!r.ok);
        assert_eq!(r.msg, "authIncorrectCreds");
        assert_eq!(r.msgi18n, Some(true));
    }

    // ── ClientMessage deserialization ────────────────────────────────────

    #[test]
    fn client_message_from_json() {
        let json = r#"{"id":1,"event":"login","args":["admin","pass"]}"#;
        let msg: ClientMessage = serde_json::from_str(json).unwrap();
        assert_eq!(msg.id, Some(1));
        assert_eq!(msg.event, "login");
        assert!(msg.args.is_some());
    }

    #[test]
    fn client_message_no_args() {
        let json = r#"{"event":"ping"}"#;
        let msg: ClientMessage = serde_json::from_str(json).unwrap();
        assert!(msg.id.is_none());
        assert_eq!(msg.event, "ping");
        assert!(msg.args.is_none());
    }
}
